File: Java/백준/Gold/30805. 사전 순 최대 공통 부분 수열/사전 순 최대 공통 부분 수열.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        int[] A = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        int m = Integer.parseInt(br.readLine());
        int[] B = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        List<Integer> result = new ArrayList<>();

        int ai = 0, bi = 0;
        while (ai < n && bi < m) {
            int maxVal = 0;
            for (int i = ai; i < n; i++) {
                for (int j = bi; j < m; j++) {
                    if (A[i] == B[j]) {
                        maxVal = Math.max(maxVal, A[i]);
                    }
                }
            }
            if (maxVal == 0) break;

            while (ai < n && A[ai] != maxVal) ai++;
            while (bi < m && B[bi] != maxVal) bi++;

            result.add(maxVal);
            ai++;
            bi++;
        }

        bw.write(result.size() + "\n");
        for (int x : result) bw.write(x + " ");
        bw.newLine();
        bw.flush();
    }
}

File: C11/백준/Gold/30805. 사전 순 최대 공통 부분 수열/사전 순 최대 공통 부분 수열.c
#include <stdio.h>

int main() {
    int n, m;
    int A[1001], B[1001];
    int result[1001];
    int ai = 0, bi = 0, rlen = 0;

    scanf("%d", &n);
    for (int i = 0; i < n; i++) scanf("%d", &A[i]);
    scanf("%d", &m);
    for (int i = 0; i < m; i++) scanf("%d", &B[i]);

    while (ai < n && bi < m) {
        int maxVal = 0;
        for (int i = ai; i < n; i++) {
            for (int j = bi; j < m; j++) {
                if (A[i] == B[j] && A[i] > maxVal) {
                    maxVal = A[i];
                }
            }
        }

        if (maxVal == 0) break;

        while (ai < n && A[ai] != maxVal) ai++;
        while (bi < m && B[bi] != maxVal) bi++;

        result[rlen++] = maxVal;
        ai++;
        bi++;
    }

    printf("%d\n", rlen);
    for (int i = 0; i < rlen; i++) {
        printf("%d ", result[i]);
    }
    printf("\n");
    return 0;
}